File: KarpRabin.py
q = 26900927
def fingerprints(s, m, d):
    global q
    t = 0
    h = (len(d)**(m-1)) % q
    for i in range(m):
        t = (len(d)*t + d[s[i]]) % q
    for i in range(len(s) - m):
        yield (s[i:i+m], t)
        t = (len(d)*(t - d[s[i]]*h) + d[s[i+m]]) % q

def fingerprint(s, d):
    global q
    p = 0
    print("len(d): {}, q: {}".format(len(d), q))
    for i in range(len(s)):
        print(p)
        p = (len(d)*p + d[s[i]]) % q
        if isinstance(p, long):
            print("Went long at {}".format(p))
    return p

def direct(s, d):
    "Directly calculate fingerprint from definition."
    global q
    b = len(d)
    return sum([n*b**m for (n, m) in zip([d[l] for l in s], reversed(range(b)))]) % q

class KarpRabin(object):
    "Uses Karp-Rabin to match a string."
    def __init__(self, alphabet):
        self.digit_values = {k:v for (k, v) in zip(alphabet, range(len(alphabet)))}
        self.q = 26900927
        self.b = len(alphabet)

    def match(self, pattern, text):
        assert len(pattern) <= len(text)
        T = [self.digit_values[k] for k in text]
        P = [self.digit_values[k] for k in pattern]
        h = self.b**(len(pattern) - 1) % self.q
        p = t = 0

        for i in range(len(pattern)):
            p = (self.b*p + P[i]) % q
            t = (self.b*t + T[i]) % q

        if (p == t): return 0

        for s in range(len(text) - len(pattern)):
            if (p == t):
                if pattern == text[s:s+len(pattern)]:
                    return s
            assert s < len(T) and s + len(pattern) < len(T)
            t = (self.b*(t - T[s]*h) + T[s+len(pattern)]) % self.q

        return None

    def __str__(self):
        return str(self.digit_values, " ", self.q, " ", self.b)

File: KarpRabin.java
/* KarpRabin.java
   A Java implementation of the Karp-Rabin seminumerical string
   matching algorithm.
   See CLRS Section 32.2.
*/
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

public class KarpRabin {
    private static final int BASE = 103;  // Arbitrary base for hash.

    /**
       Finds first match, returns true, false if no match found.
    */
    public static boolean match(String pattern, String text) {
        if (pattern.length() > text.length()) {
            return false;
        }
        int out, phash, thash;
        out = 1;

        for (int expt = pattern.length(); expt > 1; --expt) {
            out *= BASE;
        }
        
        // Calculate fingerprint of pattern and of first
        // pattern.length-length group in text.
        phash = thash = 0;
        for (int i = 0; i < pattern.length(); ++i) {
            phash = phash*BASE + pattern.charAt(i);
            thash = thash*BASE + text.charAt(i);
        }
        
        for (int s = 0; s < text.length() - pattern.length(); ++s) {
            if (phash == thash) {
                if (pattern.equals(text.substring(s, s+pattern.length()))) {
                    return true;
                }
            }
            assert s < text.length() &&
                s + pattern.length() < text.length() :
            "s is " + s + " and s+pattern.length() is " +
                (s + pattern.length());
            thash = BASE*(thash - out*text.charAt(s)) +
                text.charAt(s+pattern.length());
        }

        // See note [4].
        if (phash == thash) {
            if (pattern.equals(text.substring(text.length() - pattern.length(),
                                              text.length())))
                return true;
        }
        return false;
    }

    /**
       Finds all matches of pattern in text, returns a list of
       matching positions.
    */
    public static List<Integer> allMatches(String pattern, String text) {
        if (pattern.length() > text.length()) {
            return Collections.EMPTY_LIST;
        }
        int out, phash, thash;
        out = 1;

        for (int expt = pattern.length(); expt > 1; --expt) {
            out *= BASE;
        }
        
        // Calculate fingerprint of pattern and of first
        // pattern.length-length group in text.
        phash = thash = 0;
        for (int i = 0; i < pattern.length(); ++i) {
            phash = phash*BASE + pattern.charAt(i);
            thash = thash*BASE + text.charAt(i);
        }

        List<Integer> matches = new ArrayList<>();
        for (int s = 0; s < text.length() - pattern.length(); ++s) {
            if (phash == thash) {
                if (pattern.equals(text.substring(s, s+pattern.length()))) {
                    matches.add(s);
                }
            }
            assert s < text.length() &&
                s + pattern.length() < text.length() :
            "s is " + s + " and s+pattern.length() is " +
                (s + pattern.length());
            thash = BASE*(thash - out*text.charAt(s)) +
                text.charAt(s+pattern.length());
        }

        // See note [4].
        if (phash == thash) {
            if (pattern.equals(text.substring(text.length() - pattern.length(),
                                              text.length())))
                matches.add(text.length() - pattern.length());
        }
        if (matches.size() == 0) {
            return Collections.EMPTY_LIST;
        }
        return matches;
    }

    /* Not ready for primetime yet.
    public static Set<PatternPosition> matchSet(Set<String> patterns, String text) {
        if (pattern.length() > text.length()) {
            return Collections.EMPTY_SET;
        }
        int out, thash;
        int[] phashes = new int[pattern.size()];
        out = 1;

        for (int expt = pattern.length(); expt > 1; --expt) {
            out *= BASE;
        }
        
        thash = 0;
        for (int i = 0; i < pattern.length(); ++i) {
            phash = phash*BASE + pattern.charAt(i);
            thash = thash*BASE + text.charAt(i);
        }

        Set<PatterPosition> matches = new TreeSet<>(); 
        for (int s = 0; s < text.length() - pattern.length(); ++s) {
            if (phash == thash) {
                if (pattern.equals(text.substring(s, s+pattern.length()))) {
                    matches.add(s);
                }
            }
            assert s < text.length() &&
                s + pattern.length() < text.length() :
            "s is " + s + " and s+pattern.length() is " +
                (s + pattern.length());
            thash = BASE*(thash - out*text.charAt(s)) +
                text.charAt(s+pattern.length());
        }

        // See note [4].
        if (phash == thash) {
            if (pattern.equals(text.substring(text.length() - pattern.length(),
                                              text.length())))
                matches.add(text.length() - pattern.length());
        }
        if (matches.size() == 0) {
            return Collections.EMPTY_SET;
        }
        return matches;
        }
    */
}
